torch/distributions/kl.py

```diff
@@ -121,10 +128,211 @@ def kl_divergence(p, q):
+def _kl_exponential_exponential(p, q):
+    rate_ratio = p.rate / q.rate
+    t1 = -rate_ratio.log()
+    return t1 + rate_ratio - 1
+
 @register_kl(Gamma, Gamma)
 def _kl_gamma_gamma(p, q):
-    # Adapted from https://stats.stackexchange.com/questions/11646
-    def f(a, b, c, d):
-        return -d * a / c + b * a.log() - torch.lgamma(b) + (b - 1) * torch.digamma(d) + (1 - b) * c.log()
+    t1 = q.alpha * (p.beta / q.beta).log()
+    t2 = torch.lgamma(q.alpha) - torch.lgamma(p.alpha)
+    t3 = (p.alpha - q.alpha) * torch.digamma(p.alpha)
+    t4 = (q.beta - p.beta) * (p.alpha / p.beta)
+    return t1 + t2 + t3 + t4
+
+@register_kl(Laplace, Laplace)
+def _kl_laplace_laplace(p, q):
+    #  From http://www.mast.queensu.ca/~communications/Papers/gil-msc11.pdf
+    scale_ratio = p.scale / q.scale
+    loc_abs_diff = (p.loc - q.loc).abs()
+    t1 = -scale_ratio.log()
+    t2 = loc_abs_diff / q.scale
+    t3 = scale_ratio * torch.exp(-loc_abs_diff / p.scale)
+    return t1 + t2 + t3 - 1
+
+@register_kl(Normal, Normal)
+def _kl_normal_normal(p, q):
+    std_dev_ratio = p.std / q.std
+    t1 = -std_dev_ratio.log()
+    t2 = std_dev_ratio.pow(2)
+    t3 = ((p.mean - q.mean) / q.std).pow(2)
+    return t1 + (t2 + t3 - 1) / 2
+
+@register_kl(Pareto, Pareto)
+def _kl_pareto_pareto(p, q):
+    if p.support.lower_bound >= q.support.lower_bound:
+        #  From http://www.mast.queensu.ca/~communications/Papers/gil-msc11.pdf
+        scale_ratio = p.scale / q.scale
+        alpha_ratio = q.alpha / p.alpha
+        t1 = q.alpha * scale_ratio.log()
+        t2 = -alpha_ratio.log()
+        return t1 + t2 + alpha_ratio - 1
+    else:
+        return torch.new(p.scale.size()).fill_(float('inf'))
+
+@register_kl(Uniform, Uniform)
+def _kl_uniform_uniform(p, q):
+    if q.support.lower_bound <= p.support.lower_bound and q.support.upper_bound >= p.support.upper_bound:
+        return ((q.high - q.low) / (p.high - p.low)).log()
+    else:
+        return torch.new(q.high.size()).fill_(float('inf'))
+
+# Different distributions
+
+@register_kl(Beta, Exponential)
+def _kl_beta_exponential(p, q):
+    return -p.entropy() - q.rate.log() + q.rate * (p.alpha / (p.alpha + p.beta))
+
+@register_kl(Beta, Gamma)
+def _kl_beta_gamma(p, q):
+    t1 = -p.entropy()
+    t2 = q.alpha.lgamma() - q.alpha * q.beta.log()
+    t3 = (q.alpha - 1) * (p.alpha.digamma() - (p.alpha + p.beta).digamma())
+    t4 = q.beta * p.alpha / (p.alpha + p.beta)
+    return t1 + t2 - t3 + t4
+
+@register_kl(Beta, Normal)
+def _kl_beta_normal(p, q):
+    E_beta = p.alpha / (p.alpha + p.beta)
+    var_normal = q.std.pow(2)
+    t1 = -p.entropy()
+    t2 = 0.5 * (var_normal * 2 * math.pi).log()
+    t3 = (E_beta * (1 - E_beta) / (p.alpha + p.beta + 1) + E_beta.pow(2)) * 0.5
+    t4 = q.mean * E_beta
+    t5 = q.mean.pow(2) * 0.5
+    return t1 + t2 + (t3 - t4 + t5) / var_normal
+
+@register_kl(Beta, Pareto)
+def _kl_beta_pareto(p, q):
+    return torch.new(p.rate.size()).fill_(float('inf'))
```

Comment:
Consider adding a helper
```py
def _infinity_like(tensor):
    return tensor.new([float('inf')]).expand_as(tensor)
```
This version also has correct GPU placement and is cheaper than `fill_`.